server/src-lib/Data/Time/Clock/Units.hs

```diff
@@ -69,56 +72,276 @@ import           Prelude
+
+deriving via (TimeUnit 1000000 'Absolute) instance Fractional (Microseconds 'Absolute)
+deriving via (TimeUnit 1000000 'Calendar) instance Fractional (Microseconds 'Calendar)
+
+deriving via (TimeUnit 1000000 'Absolute) instance Real (Microseconds 'Absolute)
+deriving via (TimeUnit 1000000 'Calendar) instance Real (Microseconds 'Calendar)
+
+deriving via (TimeUnit 1000000 'Absolute) instance Hashable (Microseconds 'Absolute)
+deriving via (TimeUnit 1000000 'Calendar) instance Hashable (Microseconds 'Calendar)
+
+deriving via (TimeUnit 1000000 'Absolute) instance RealFrac (Microseconds 'Absolute)
+deriving via (TimeUnit 1000000 'Calendar) instance RealFrac (Microseconds 'Calendar)
+
+
+newtype Nanoseconds t = Nanoseconds { nanoseconds :: Duration t}
+
+deriving instance Duration' (Nanoseconds 'Absolute)
+deriving instance Duration' (Nanoseconds 'Calendar)
+
+deriving instance Show (Nanoseconds 'Absolute)
+deriving instance Show (Nanoseconds 'Calendar)
+
+deriving instance Eq (Nanoseconds 'Absolute)
+deriving instance Eq (Nanoseconds 'Calendar)
+
+deriving instance Ord (Nanoseconds 'Absolute)
+deriving instance Ord (Nanoseconds 'Calendar)
+
+deriving via (TimeUnit 1000 'Absolute) instance Read (Nanoseconds 'Absolute)
+deriving via (TimeUnit 1000 'Calendar) instance Read (Nanoseconds 'Calendar)
+
+deriving via (TimeUnit 1000 'Absolute) instance Num (Nanoseconds 'Absolute)
+deriving via (TimeUnit 1000 'Calendar) instance Num (Nanoseconds 'Calendar)
+
+deriving via (TimeUnit 1000 'Absolute) instance Fractional (Nanoseconds 'Absolute)
+deriving via (TimeUnit 1000 'Calendar) instance Fractional (Nanoseconds 'Calendar)
+
+deriving via (TimeUnit 1000 'Absolute) instance Real (Nanoseconds 'Absolute)
+deriving via (TimeUnit 1000 'Calendar) instance Real (Nanoseconds 'Calendar)
+
+deriving via (TimeUnit 1000 'Absolute) instance Hashable (Nanoseconds 'Absolute)
+deriving via (TimeUnit 1000 'Calendar) instance Hashable (Nanoseconds 'Calendar)
+
+deriving via (TimeUnit 1000 'Absolute) instance RealFrac (Nanoseconds 'Absolute)
+deriving via (TimeUnit 1000 'Calendar) instance RealFrac (Nanoseconds 'Calendar)
 
 -- Internal for deriving via
-newtype TimeUnit (picosPerUnit :: Nat) = TimeUnit DiffTime
-  deriving (Show, Eq, Ord)
+newtype TimeUnit (picosPerUnit :: Nat) t = TimeUnit (Duration t)
+
+deriving instance Show (TimeUnit picosPerUnit 'Absolute)
+deriving instance Show (TimeUnit picosPerUnit 'Calendar)
+
+deriving instance Eq (TimeUnit picosPerUnit 'Absolute)
+deriving instance Eq (TimeUnit picosPerUnit 'Calendar)
+
+deriving instance Ord (TimeUnit picosPerUnit 'Absolute)
+deriving instance Ord (TimeUnit picosPerUnit 'Calendar)
 
 type SecondsP n = n GHC.TypeLits.* 1000000000000
 
 natNum :: forall n a. (KnownNat n, Num a) => a
 natNum = fromInteger $ natVal (Proxy @n)
 
-instance (KnownNat picosPerUnit) => Num (TimeUnit picosPerUnit) where
+instance (KnownNat picosPerUnit) => Num (TimeUnit picosPerUnit 'Absolute) where
```

Comment:
It should be possible to avoid duplicating all these instances between `Absolute` and `Calendar`. The main obstacle here is that we need generic versions of `picosecondsToDiffTime` and `diffTimeToPicoseconds`. Let’s define a helper class:

```haskell
class AsPicoseconds t where
  toPicoseconds :: Duration t -> Pico
  fromPicoseconds :: Pico -> Duration t

instance AsPicoseconds 'Absolute where
  toPicoseconds = diffTimeToPicoseconds
  fromPicoseconds = picosecondsToDiffTime

instance AsPicoseconds 'Calendar where
  toPicoseconds = nominalDiffTimeToPicoseconds
  fromPicoseconds = picosecondsToNominalDiffTime
```

Now you can change this instance declaration to

```haskell
instance (KnownNat picosPerUnit, AsPicoseconds t) => Num (TimeUnit picosPerUnit t) where
```

and use the generic functions in the body, and it should just work. Likewise, all the standalone derived instances can be rephrased as

```haskell
deriving instance AsPicoseconds t => Num (Nanoseconds t)
```

and so on.